DDDEastAnglia/Helpers/LoginMethods/ExternalLoginsProvider.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using DDDEastAnglia.Helpers.Security;
+using DDDEastAnglia.Models;
+using Microsoft.Web.WebPages.OAuth;
+
+namespace DDDEastAnglia.Helpers.LoginMethods
+{
+    public class ExternalLoginsProvider
+    {
+        private readonly IOAuthClientInfo oAuthClientInfo;
+
+        public ExternalLoginsProvider(IOAuthClientInfo oAuthClientInfo)
+        {
+            if (oAuthClientInfo == null)
+            {
+                throw new ArgumentNullException("oAuthClientInfo");
+            }
+            
+            this.oAuthClientInfo = oAuthClientInfo;
+        }
+
+        public IEnumerable<LoginMethod> GetForUser(string userName)
+        {
+            var accounts = oAuthClientInfo.GetAccountsFromUserName(userName);
+            var externalLogins = new List<LoginMethod>();
+
+            foreach (OAuthAccount account in accounts)
+            {
+                var clientData = oAuthClientInfo.GetOAuthClientData(account.Provider);
+
+                externalLogins.Add(new LoginMethod
+                {
+                    DisplayName = clientData.DisplayName,
+                    ProviderName = account.Provider,
+                    ProviderUserId = account.ProviderUserId
+                });
+            }
+
+            return externalLogins.OrderBy(l => l.DisplayName);
+        }
+
+        public IEnumerable<LoginMethod> GetAllAvailable()
+        {
+            var authenticationClientDatas = oAuthClientInfo.RegisteredClientData();
+            var externalLogins = new List<LoginMethod>();
+
+            foreach (var externalLogin in authenticationClientDatas)
+            {
+                var provider = externalLogin.AuthenticationClient.ProviderName;
+
+                externalLogins.Add(new LoginMethod
+                {
+                    DisplayName = externalLogin.DisplayName,
+                    ProviderName = provider
+                });
+            }
+
+            return externalLogins.OrderBy(l => l.DisplayName);
```

Comment:
Is it worth doing the ordering up-front on the `foreach`, and then `yield return`ing each element from the method, rather than building a list and returning that? Given the size of the list, perhaps not, but worth a thought at least.